File: lib/src/exceptions/invalid_musicxml_exception.dart
/// Exception thrown when an invalid MusicXML file is encountered.
class InvalidMusicXmlException implements Exception {
  /// The error message.
  final String message;

  /// The node where the error occurred, if available.
  final String? node;

  /// The line number where the error occurred, if available.
  final int? line;

  /// Creates a new [InvalidMusicXmlException] with the given [message].
  InvalidMusicXmlException(this.message, {this.node, this.line});

  @override
  String toString() {
    final buffer = StringBuffer('InvalidMusicXmlException: $message');
    if (node != null) {
      buffer.write(' (node: $node');
      if (line != null) {
        buffer.write(', line: $line');
      }
      buffer.write(')');
    } else if (line != null) {
      buffer.write(' (line: $line)');
    }
    return buffer.toString();
  }
}

File: csharp/MusicXMLParser/Exceptions/InvalidMusicXmlException.cs
using System;
using System.Text;

namespace MusicXMLParser.Exceptions
{
    /// <summary>
    /// Base exception for errors encountered during MusicXML parsing or validation.
    /// </summary>
    public class InvalidMusicXmlException : Exception
    {
        /// <summary>
        /// The XML node content or identifier where the error occurred, if available.
        /// </summary>
        public string? Node { get; }

        /// <summary>
        /// The line number in the XML file where the error occurred, if available.
        /// </summary>
        public string? Line { get; }

        /// <summary>
        /// Creates a new <see cref="InvalidMusicXmlException"/> with the given message.
        /// </summary>
        /// <param name="message">The error message.</param>
        /// <param name="line">The line number where the error occurred (optional).</param>
        /// <param name="node">The XML node where the error occurred (optional).</param>
        public InvalidMusicXmlException(string message, string? line = null, string? node = null)
            : base(message)
        {
            Node = node;
            Line = line;
        }

        /// <summary>
        /// Creates a new <see cref="InvalidMusicXmlException"/> with the given message and inner exception.
        /// </summary>
        /// <param name="message">The error message.</param>
        /// <param name="innerException">The inner exception.</param>
        /// <param name="line">The line number where the error occurred (optional).</param>
        /// <param name="node">The XML node where the error occurred (optional).</param>
        public InvalidMusicXmlException(string message, Exception? innerException, string? line = null, string? node = null)
            : base(message, innerException)
        {
            Node = node;
            Line = line;
        }

        public override string ToString()
        {
            var buffer = new StringBuilder($"{GetType().Name}: {Message}"); // Use GetType().Name for specific exception type
            if (!string.IsNullOrEmpty(Node))
            {
                buffer.Append($" (node: {Node}");
                if (!string.IsNullOrEmpty(Line))
                {
                    buffer.Append($", line: {Line}");
                }
                buffer.Append(")");
            }
            else if (!string.IsNullOrEmpty(Line))
            {
                buffer.Append($" (line: {Line})");
            }
            return buffer.ToString();
        }
    }
}
